Tryouts/Plugins/Services/MockAuthentication/MockAuthenticationService.cs

```diff
@@ -0,0 +1,75 @@
+﻿/// ********************************************************************************************************
+///
+/// Morgan Stanley makes this available to you under the Apache License, Version 2.0 (the "License").
+/// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+/// See the NOTICE file distributed with this work for additional information regarding copyright ownership.
+/// Unless required by applicable law or agreed to in writing, software distributed under the License
+/// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and limitations under the License.
+/// 
+/// ********************************************************************************************************
+
+using NP.Utilities;
+using NP.Utilities.Attributes;
+using NP.Utilities.BasicServices;
+
+namespace MockAuthentication;
+
+[Implements(typeof(IAuthenticationService), IsSingleton = true)]
```

Comment:
This is being discussed by the team. There was a heated debate under the old (now closed) PR and on a few meetings. Personally I'm strongly in favor of using standard, widely adopted techniques like MSFT DI and logging abstractions, at least in the public APIs. Also I'm strongly against implicit wiring with magic attributes etc.

Reply:
We discussed it yes, pretty heatedly. I still think that Microsofts DependencyInjection package lacks major functionality and even if we do not add it now - we'll have to add it in  the future because of the client requests. 
The attributes are required for dynamically loading the plugins though we can discuss improving the attribute names.

Reply:
Right, I guess we're having DI/IoC conversations here 😝 

@BalassaMarton made 2 points
1. I'm strongly in favor of using standard, widely adopted techniques like MSFT DI and logging **abstractions**, at least in the public APIs. 
2. I'm strongly against implicit wiring with magic attributes etc

@nick-polyak-ms your response was 
1.  I still think that Microsofts DependencyInjection package lacks major functionality and even if we do not add it now - we'll have to add it in the future because of the client requests.
2. The attributes are required for dynamically loading the plugins though we can discuss improving the attribute names.

**Regarding 1.** 
@BalassaMarton explicitly stated abstractions, @nick-polyak-ms you are talking about implementations. IMO these are 2 separate topics.  e.g. it is possible to use IServiceCollection/IServiceProvider/IServiceScopeFactory/IServiceScope and provide an implementation that isn't ServiceCollection based but is instead an adaptation of another container to these abstractions.  The same can be said for taking a dependency on different abstractions and being able to swap implementations.  A real life example in the space of Composite UI Application Frameworks (CUIAF) is Prism.  Prism defines their own set of abstractions, which you can view [here](https://github.com/PrismLibrary/Prism/tree/master/src/Prism.Core/Ioc).  Prism internals only ever depend on these.  You can put together a Prism application that uses Unity or DryIoc with packages provided by Prism.  You can also put together a Prism application using another container as long as it satisfies the requirements that Prism puts forward.  Because of the implementation details of the Prism bootstrapping process, one of these requirements is that the container implementation must be mutable.  That is, you must be able to register, then resolve, then register more, then resolve, ad infinitum.  Because of this, you can not use immutable implementations like ServicesCollection (I'm talking about Microsoft DI implementation, not abstraction here).  Note that I think changes to the Prism IoC abstractions and the bootstrapping process could loosen the requirement, but that using immutable implementations would perform more poorly in the bootstrapping path.  If this is acceptable to an application builder, then they could just stick with ServicesCollection.

Separate from the discussion of core abstractions, if there are core parts of container configuration that rely on specific container implementation features, then that can be done in a centralised location by taking a dependency on that implementation.  Elsewhere all programmatic usage can be dependent on the abstractions.  Note however, that if this is done then there is a hard dependency taken on an implementation.  If this is done in the core of a CUIAF, then you force your consumers to use this container, with no choice.

**Regarding 2.** 
If a decision is to be made around usage or not of container auto-wiring, then I think the conversation leading to the decision needs to cover at least:
* What are the details of perceived pros/cons (in general)
* Are there scenarios that require the feature, and if so the details of why it's required need to be understood by all and documented.
With the above we can do a cost/benefit analysis that feeds into the ultimate decision.  @BalassaMarton instead of just saying your against it, tell us the details of what you see as the drawbacks.  @nick-polyak-ms tell us the details of what value you think it provides.  Also if it is required for a specific Plugin model, elaborate on the details of the scenario and illustrate how the model doesn't work without it.  Use code if you have to.

Related to this, there is discussion of Plugins.  I don't see the notion of Plugin defined anywhere.  It's not clear to me what Plugin even is in this context.  That makes it hard for me to imagine why it might be required for Plugins.

Final point.  **If** container auto-configuration is indeed required by core scenarios, **then** there is a separate discussion of whether this is done by Attributes or by convention.  That would also require a cost/benefit analysis.  I'll add some input here that I would feed into that discussion.
Personal Opinions:
* I value concise code over verbose code
Notes:
* In terms of implementation of auto-wiring.  Both Attribute based and convention based can be implemented on top of abstractions and don't require it to be baked into the Container implementation
* Attribute based auto-wiring logic can be directly implemented in/on a Container.  It can also be implemented as one specific type of Convention for an implementation that already supports Conventions.
Observations feeding into perceived Pros/Cons
* If the conventions you will follow aren't already implemented for a given container, then work is required to implement those conventions in code.
* The attribute model is much more verbose in files that have implementations of abstractions.
* The attribute model is much more verbose in files that have implementation of Types that follow IoC and therefore expect dependencies to be injected
* Everything expressed in the Attributes either already is expressed by the existing Type system, or can be additionally expressed through naming conventions.  For example
```
[Implements(typeof(IAuthenticationService), IsSingleton = true)]
public class AuthenticationService : IAuthenticationService
```
Can also be expressed as follows and Convention based auto-configuration can still do the same configuration
```
public class SingletonAuthenticationService : IAuthenticationService
```

Reply:
I need an implementation of a simple interface that allow adding and resolving by (type, name) pair. Part of the reason, is that shell does not need to know the exact type of dynamically loaded plugin - it can resolve it say by IPlugin interface plus some name unique for that plugin. 

The attributes are better to keep, because otherwise, it means that EVERY type needs to be dynamically injected into a container - which is usually not the case.

Reply:
We can use any container that has name resolution to implement that adapter. 
Microsoft failed with some other products and APIs (WindowsRT, UWP etc, etc) and their poor container is just an example of failure to me.